rfcs/2021-07-19-8216-multiple-pipelines.md

```diff
@@ -0,0 +1,260 @@
+
+```rust
+pub trait FunctionTransform: ... {
+  // returning the name and the version
+  fn pipeline(&self) -> Option<(String, String)>;
+  // name of the transform
+  fn name(&self) -> String;
+}
+```
+
+And then adding a `pre_transform` function that would add the pipeline fields
+
+```rust
+pub trait FunctionTransform: ... {
+  fn transform(&mut self, output: &mut Vec<Event>, event: Event) {
+    self.do_transform(output, self.pre_transform);
+  }
+
+  fn pre_transform(&mut self, event: Event) -> Event {
+    // update the event in order to add the field
+  }
+
+  fn transform(&mut self, output: &mut Vec<Event>, event: Event);
+}
+```
+
+This solution would required the core of vector and how it handles transforms but would avoid generatic dynamic transforms.
+It would probably also reduce the performance considering that each transform would have to check if it's part of a pipeline.
+
+## Doc-level Proposal
+
+With this pipeline definition, the following `vector.toml`
+
+```toml
+[sources.docker]
+type = "docker_logs"
+...
+
+[transforms.nginx_filter]
+input = ["docker"]
+type = "remap"
+source = '''
+  # do something that filter nginx logs from docker logs
+'''
+
+[transforms.nginx]
+input = ["nginx_filter"]
+type = "remap"
+source = '''
+  # do some transformations with the nginx logs
+'''
+
+[transforms.apache_filter]
+input = ["docker"]
+type = "remap"
+source = '''
+  # do something that filter apache logs from docker logs
+'''
+
+[transforms.apache]
+input = ["apache_filter"]
+type = "remap"
+source = '''
+  # do some transformations with the apache logs
+'''
+
+[sinks.ouput]
+inputs = ["apache", "nginx"]
+type = "console"
+...
+```
+
+can be split into the following `pipelines/frontend.toml`
+
+```toml
+[transforms.nginx_filter]
+input = ["docker"]
+type = "remap"
+source = '''
+  # do something that filter nginx logs from docker logs
+'''
+
+[transforms.nginx]
+input = ["nginx_filter"]
+type = "remap"
+source = '''
+  # do some transformations with the nginx logs
+'''
+
+[transforms.apache_filter]
+input = ["docker"]
+type = "remap"
+source = '''
+  # do something that filter apache logs from docker logs
+'''
+
+[transforms.apache]
+input = ["apache_filter"]
+type = "remap"
+source = '''
+  # do some transformations with the apache logs
+'''
+```
+
+and `vector.toml`
+
+```toml
+[sources.docker]
+type = "docker_logs"
+...
+
+[sinks.ouput]
+inputs = ["frontend.apache", "frontend.nginx"]
+type = "console"
+...
+```
+
+## Rationale
+
+This improves the readability of the `vector.toml` and add the possibility to share pipelines or deploy them with some tools like Terraform or Ansible without modifying the main configuration.
+In the future, this will allow us to replicate datadog's pipelines locally.
+
+## Drawbacks
+
+- Not yet able to use a pipeline in another pipeline, but this may come in further work.
+- First option: creating some hidden transforms and modifying the existing ones could introduce some bugs.
+
+## Alternatives
```

Comment:
An RFC with a low number of alternatives concerns me. It's fine to have a position you're arguing for -- that's expected -- but with so little else considered it doesn't strike me that we've thought through the approach stumped for here in detail compared to other approaches. Why not, for instance:

* do nothing (are users asking for this capability directly?)
* evolve vector to use a tag/filter model like our competitors, have a 'pipeline' be a 'tag'
* run a single vector per-'pipeline' and support metric tagging to distinguish at the telemetry level 

Etc etc. I'm not stumping for any of these but I'd like to see more consideration of different ways of approaching this before I have confidence that the one being proposed in this RFC is the best, cleanest way forward.

Reply:
Agree! This is partly my fault since the brief was opinionated on this approach. @zsherman and I have decided to adjust the product briefs to better consider alternatives. That said, I would not mind walking through some of these alternatives to see where we land.

Reply:
To be honest, I don't think I've a deep enough understanding of Vector (and from our competitors) to tell what would be the pros/cons of those alternatives 🤔